File: src/test/java/im/tox/gui/FriendList.java
package im.tox.gui;

import im.tox.tox4j.core.enums.ToxConnection;
import im.tox.tox4j.core.enums.ToxStatus;

import javax.swing.*;
import java.util.ArrayList;
import java.util.Arrays;

public class FriendList extends AbstractListModel<Friend> {

    private final ArrayList<Friend> friends = new ArrayList<>();

    public void add(int friendNumber, byte[] publicKey) {
        while (friends.size() <= friendNumber) {
            friends.add(null);
        }
        Friend oldFriend = friends.get(friendNumber);
        if (oldFriend == null || !Arrays.equals(oldFriend.getPublicKey(), publicKey)) {
            friends.set(friendNumber, new Friend(publicKey));
        }
        fireIntervalAdded(this, friendNumber, friendNumber);
    }

    @Override
    public int getSize() {
        return friends.size();
    }

    @Override
    public Friend getElementAt(int index) {
        return friends.get(index);
    }

    public void setName(int friendNumber, String name) {
        friends.get(friendNumber).setName(name);
        fireContentsChanged(this, friendNumber, friendNumber);
    }

    public void setConnectionStatus(int friendNumber, ToxConnection connectionStatus) {
        friends.get(friendNumber).setConnectionStatus(connectionStatus);
        fireContentsChanged(this, friendNumber, friendNumber);
    }

    public void setStatus(int friendNumber, ToxStatus status) {
        friends.get(friendNumber).setStatus(status);
        fireContentsChanged(this, friendNumber, friendNumber);
    }

    public void setStatusMessage(int friendNumber, String message) {
        friends.get(friendNumber).setStatusMessage(message);
        fireContentsChanged(this, friendNumber, friendNumber);
    }

    public void setTyping(int friendNumber, boolean isTyping) {
        friends.get(friendNumber).setTyping(isTyping);
        fireContentsChanged(this, friendNumber, friendNumber);
    }

}

File: projects/tox4j/src/test/java/im/tox/gui/domain/FriendList.scala
package im.tox.gui.domain

import javax.swing._

import im.tox.tox4j.core.enums.{ ToxConnection, ToxUserStatus }

import scala.collection.mutable.ArrayBuffer

final class FriendList extends AbstractListModel[Friend] {
  private val friends = new ArrayBuffer[Friend]

  /**
   * Add a friend to the friend list with the associated public key.
   *
   * @param friendNumber Friend number from toxcore.
   * @param publicKey Public key as stable identifier for the friend.
   */
  def add(friendNumber: Int, publicKey: Array[Byte]): Unit = {
    while (friends.size <= friendNumber) {
      friends += null
    }

    val oldFriend = friends(friendNumber)
    if (oldFriend == null || oldFriend.publicKey.deep != publicKey.deep) {
      friends(friendNumber) = new Friend(publicKey)
    }

    fireIntervalAdded(this, friendNumber, friendNumber)
  }

  override def getSize: Int = friends.size
  override def getElementAt(index: Int): Friend = friends(index)

  def setName(friendNumber: Int, name: String): Unit = {
    friends(friendNumber).name = name
    fireContentsChanged(this, friendNumber, friendNumber)
  }

  def setConnectionStatus(friendNumber: Int, connectionStatus: ToxConnection): Unit = {
    friends(friendNumber).connectionStatus = connectionStatus
    fireContentsChanged(this, friendNumber, friendNumber)
  }

  def setStatus(friendNumber: Int, status: ToxUserStatus): Unit = {
    friends(friendNumber).status = status
    fireContentsChanged(this, friendNumber, friendNumber)
  }

  def setStatusMessage(friendNumber: Int, message: String): Unit = {
    friends(friendNumber).statusMessage = message
    fireContentsChanged(this, friendNumber, friendNumber)
  }

  def setTyping(friendNumber: Int, isTyping: Boolean): Unit = {
    friends(friendNumber).typing = isTyping
    fireContentsChanged(this, friendNumber, friendNumber)
  }
}
